flink-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KillerWatchDog.java

```diff
@@ -42,12 +42,12 @@
 	@SuppressWarnings("deprecation")
 	@Override
 	public void run() {
-		final long deadline = System.currentTimeMillis() + timeout;
+		final long deadline = System.nanoTime() + timeout * 1_000_000L;
 		long now;
 
-		while (toKill.isAlive() && (now = System.currentTimeMillis()) < deadline) {
+		while (toKill.isAlive() && (now = System.nanoTime()) < deadline) {
 			try {
-				toKill.join(deadline - now);
+				toKill.join((deadline - now) / 1_000_000L);
```

Comment:
thanks for pointing this, my mistake for ignoring the API of Thread#join:) I proposed a simple move of rounding.

It is found that the Thread#join is using the System.currentTimeMillis(). So the fix may be not  as strong as we expected. The better may be that we do not fix KillerWatchDog because the "last resort" codes already there.